src/event/vrl_target.rs

```diff
@@ -0,0 +1,398 @@
+use super::{Event, Metric, MetricKind, Value};
+use snafu::Snafu;
+use std::{collections::BTreeMap, convert::TryFrom, iter::FromIterator};
+use vrl::path::Segment;
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set  it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// An adapter to turn `Event`s into `vr::Target`s
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
+    LogEvent(vrl::Value),
+    Metric(Metric),
+}
+
+impl VrlTarget {
+    pub fn new(event: Event) -> Self {
+        match event {
+            // TODO optimize
+            Event::Log(mut event) => VrlTarget::LogEvent(vrl::Value::Object(
+                event
+                    .take_fields()
+                    .into_iter()
+                    .map(|(key, value)| (key, value.into()))
+                    .collect(),
+            )),
+            Event::Metric(event) => VrlTarget::Metric(event),
+        }
+    }
+
+    /// Turn the target back into events
+    ///
+    /// This returns an iterator of events as one event can be turned into multiple by assign `.`
+    /// to an array in VRL
+    pub fn into_events(self) -> impl Iterator<Item = Event> {
+        match self {
+            VrlTarget::LogEvent(value) => {
+                Box::new(value_into_events(value)) as Box<dyn Iterator<Item = Event>>
```

Comment:
Is the explicit `Box<dyn Iterator<Item = Event>>` needed because the compiler chooses a too specific type for each arm?

Reply:
Yeah, that's exactly it :)

Reply:
`as Box<dyn _>` returns a compiler error:

```
error: `<` is interpreted as a start of generic arguments for `Box`, not a comparison
  --> src/event/vrl_target.rs:48:72
   |
48 |                 Box::new(std::iter::once(Event::Metric(metric))) as Box<dyn _>
   |                 -------------------------------------------------------^----- interpreted as generic arguments
   |                 |                                                      |
   |                 |                                                      not interpreted as comparison
   |                 help: try comparing the cast value: `(Box::new(std::iter::once(Event::Metric(metric))) as Box)`

error: expected expression, found keyword `dyn`
  --> src/event/vrl_target.rs:48:73
   |
48 |                 Box::new(std::iter::once(Event::Metric(metric))) as Box<dyn _>
   |                                                                         ^^^ expected expression
```